<a href="https://colab.research.google.com/github/CarlaFFochs/Udemy_Projects/blob/main/0_Typical_Workflow_Sklearn_(Heart_disease).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##An end-to-end Sckit-Learn workflow

Check the whole file --> https://github.com/mrdbourke/zero-to-mastery-ml/blob/master/section-2-data-science-and-ml-tools/introduction-to-scikit-learn.ipynb

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
import numpy as np
heart_disease = pd.read_csv("/content/drive/MyDrive/MASTER DATA SCIENCE/M0/M0 - UDEMY/Machine Learning/Typical Workflow Sklearn/data/heart-disease.csv")
heart_disease.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [3]:
#Create X (feautre matrix)
X = heart_disease.drop("target", axis=1)

#Create y (target)
y = heart_disease["target"]

In [4]:
X.head(1)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   

   ca  thal  
0   0     1

In [5]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [6]:
#2.Choose the right model and hyperparameters

from sklearn.ensemble import RandomForestClassifier
clf= RandomForestClassifier()

#We'll keep the default hyperparametrs
clf.get_params() #what parameters your model is using

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

In [7]:
#3. Fit the model to the data
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y , test_size = 0.2)

#80% of the data will be used for traning
#20& of the data will ne used fot test


In [8]:
clf.fit(X_train, y_train) #our model has been fit (ja el tenim entrenat)

RandomForestClassifier()

In [9]:
#make a prediciton

#y_label= clf.predict(np.array([0 , 1 , 4 , 0]))

#we can only make predictions with the same shape of the X and the above array has
#not the same shape

In [10]:
X.shape #we have 303 rows and 13 columns

(303, 13)

In [11]:
X_test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
180   55    1   0       132   353    0        1      132      1      1.2   
122   41    0   2       112   268    0        0      172      1      0.0   
271   61    1   3       134   234    0        1      145      0      2.6   
29    53    1   2       130   197    1        0      152      0      1.2   
223   56    0   0       200   288    1        0      133      1      4.0   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
302   57    0   1       130   236    0        0      174      0      0.0   
151   71    0   0       112   149    0        1      125      0      1.6   
252   62    0   0       138   294    1        1      106      0      1.9   
43    53    0   0       130   264    0        0      143      0      0.4   
27    51    1   2       110   175    0        1      123      0      0.6   

     slope  ca  thal  
180      1   1     3  
122      2   0     2  
271      1   2     2  
29       0   0     2  
223      0   2     3  
..     ...  ..   ...  
302      1   1     2  
151      1   0     2  
252      1   3     2  
43       1   0     2  
27       2   0     2  

[61 rows x 13 columns]

In [12]:
y_test

180    0
122    1
271    0
29     1
223    0
      ..
302    0
151    1
252    0
43     1
27     1
Name: target, Length: 61, dtype: int64

In [13]:
y_preds= clf.predict(X_test) #sempre fem prediccions amb el valor del TEST, amb el TRAIN no té sentit

In [14]:
y_preds

array([0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1,
       0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1])

In [15]:
#4 . Evaluate the model on the training data and test data

clf.score(X_train,y_train)

1.0

In [16]:
clf.score(X_test, y_test) #accuracy of the test data

0.8360655737704918

In [17]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       0.86      0.80      0.83        30
           1       0.82      0.87      0.84        31

    accuracy                           0.84        61
   macro avg       0.84      0.84      0.84        61
weighted avg       0.84      0.84      0.84        61



In [18]:
confusion_matrix(y_test, y_preds) 

array([[24,  6],
       [ 4, 27]])

In [19]:
list1= ["TN","FP"]
list2= ["FP", "TP"]
list1


['TN', 'FP']

In [20]:
list2

['FP', 'TP']

In [21]:
matrix= np.array([[list1],[list2]])
matrix

array([[['TN', 'FP']],

       [['FP', 'TP']]], dtype='<U2')

In [22]:
confusion_matrix(y_test, y_preds) 
 #hi han 

array([[24,  6],
       [ 4, 27]])

A confusion matrix is a table that is often used to **describe the performance of a classification model** (or "classifier") on a set of test data for which the true values are known. The confusion matrix itself is relatively simple to understand, but the related terminology can be confusing.

Let's start with an example confusion matrix for a binary classifier (though it can easily be extended to the case of more than two classes):

![title](https://www.dataschool.io/content/images/2015/01/confusion_matrix_simple2.png)


**What can we learn from this matrix?**

There are two possible predicted classes: "yes" and "no". If we were predicting the presence of a disease, for example, "yes" would mean they have the disease, and "no" would mean they don't have the disease.
The classifier made a total of 165 predictions (e.g., 165 patients were being tested for the presence of that disease).



> **Out of those 165 cases, the classifier predicted "yes" 110 times, and "no" 55 times.**
**In reality, 105 patients in the sample have the disease, and 60 patients do not**.

Let's now define the most basic terms, which are whole numbers (not rates):



*   **True Positives (TP):** These are cases in which we **predicted YES** (they have the disease), and they **do have the disease**.
*   **True Negatives (TN):** We **predicted NO**, and they **don't have the disease.**

*   **False Positives (FP):** We **predicted YES**, but they **don't actually have the disease.** (Also known as a "Type I error.")
*   **False Negatives (FN):** We **predicted NO**, but they **actually do have the disease.** (Also known as a "Type II error.")


I've added these terms to the confusion matrix, and also added the row and column totals:

![title](https://www.dataschool.io/content/images/2015/01/confusion_matrix2.png)

In [23]:
accuracy_score(y_test, y_preds)

0.8360655737704918

In [24]:
#5. Improve a model
#Try different amount of n_estimators

np.random.seed(42)
for i in range (10, 100, 10): #el últim valor es de 10
  print(f"Trying model with {i} estimators...")
  clf = RandomForestClassifier(n_estimators=i).fit(X_train, y_train)
  print(f"Model accuracy on test set: {clf.score(X_test, y_test) * 100:.2f}")
  print(" ")

Trying model with 10 estimators...
Model accuracy on test set: 78.69
 
Trying model with 20 estimators...
Model accuracy on test set: 86.89
 
Trying model with 30 estimators...
Model accuracy on test set: 78.69
 
Trying model with 40 estimators...
Model accuracy on test set: 83.61
 
Trying model with 50 estimators...
Model accuracy on test set: 81.97
 
Trying model with 60 estimators...
Model accuracy on test set: 80.33
 
Trying model with 70 estimators...
Model accuracy on test set: 81.97
 
Trying model with 80 estimators...
Model accuracy on test set: 78.69
 
Trying model with 90 estimators...
Model accuracy on test set: 81.97
 


In [25]:
#6. Save a model and load it

import pickle

pickle.dump(clf, open("random_forst_model_1.pkl", "wb")) #write binary

#t'apareix a la teva notebook de jupiter

In [26]:
loaded_model= pickle.load(open("random_forst_model_1.pkl", "rb")) #read binary
loaded_model.score(X_test, y_test)

0.819672131147541

**Accuracy:** Overall, how often is the classifier correct?
- (TP+TN)/total = (100+50)/165 = 0.91

**Misclassification Rate:** Overall, how often is it wrong?
- FP+FN)/total = (10+5)/165 = 0.09
- equivalent to 1 minus Accuracy
- also known as "Error Rate"

**True Positive Rate:** When it's actually yes, how often does it predict yes?
- TP/actual yes = 100/105 = 0.95
- also known as "Sensitivity" or "Recall"

**False Positive Rate:** When it's actually no, how often does it predict yes?
- FP/actual no = 10/60 = 0.17

**True Negative Rate:** When it's actually no, how often does it predict no?
- TN/actual no = 50/60 = 0.83
- equivalent to 1 minus False Positive Rate
- also known as "Specificity"

**Precision:** When it predicts yes, how often is it correct?
- TP/predicted yes = 100/110 = 0.91

**Prevalence:** How often does the yes condition actually occur in our sample?
- actual yes/total = 105/165 = 0.64